---
title: "MSR and expression for heart left ventricle, windows of 1000 sites"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#knitr::opts_chunk$set(fig.width=20, fig.height=12) 

#setwd(dir = "Scrivania/Tesi/Methylation-project/Code/")
#source("Methylation-project/Code/expression_and_msr_experiment.R", chdir = T)
```



```{r, include=FALSE}
library("PerformanceAnalytics")
setwd(dir = "../../../")
source("Methylation-project/Code/WGBS_analysis_functions.R", chdir = T)


##########################
genebody = readRDS("Rexperiments/genebody_improved.Rda")
##########################

y <- c("gene_count", "log_tpm")
basic_predictors <- c("nucleotides", "CpG_density", "meth rate", "msr_density")
essentials <- c(y, basic_predictors)
essential_msr_predictors <- c("msr", "inverted_msr")
fancy_msr_predictors <- c("ecdf", "inverted ecdf", "residual", "inverted_residual")
msr_predictors <- c(essential_msr_predictors, fancy_msr_predictors)
```


```{r, include=FALSE}
setwd(dir = "../../../")
rna <- read_rna_file("MethylationCode/MethylationData/rna-seq/ENCFF860DPP_left_ventricle_male_34.tsv")
msr_fragments_table = readRDS("Rexperiments/ENCFF536RSX_left_ventricle_msr_table_1000.Rda")
rna_fragment_table = readRDS("Rexperiments/ENCFF536RSX_left_ventricle_rna_table_1000.Rda")
#########################
```

Here I will investigate if there is a relationship between the presence of genes and genes expression in a certain genomic region and the MSR (with some of its derivate statistics).

I chose heart left ventricle cells data in order to do this, and CpG windows of size 1000, that corresponds to a variable window size in term of nucleotides (on average about 100,000).


This is an example of total-rna-seq file, that shows for each "gene" its transcripts and some measures of expression. In this case I just kept two colums.
The first one indicates the "gene"", the second one is the Transcript Per Million that is a relative measure of how much a gene is expressed.

```{r, echo=FALSE}
#heart left ventricle_rna
rna[nchar(gene_id)>6]
```

This is the annotation file that store the position occupied by each human gene.

```{r, echo=FALSE}
genebody[,c(1,2,3,4,5,6)]
```
The number of genes is much less than the ones in the total-rna-seq file, since the first one also contains so called pseudogenes and other stuff.



So the final dataFrame is the following (excluding some columns for readability):
```{r, echo=FALSE}
data_table <- join_rna_and_msr_table(rna_fragment_table, msr_fragments_table)

data_table <- exclude_outliers(data_table, lim = 2.5e5)
data_table$log_tpm[data_table$gene_count==0] = NA
#data_table <- exclude_outliers(data_table, lim = 2e6)
head(data_table[complete.cases(data_table), c(1,2,3,5,7,11)])
```

The full scheme includes:

**nucleotides**: number of nucleotides in the window

**CpG density**: fraction of nucleotides that is a C of a CpG site (= 1000/nucleotides)

**meth rate**: ratio of methylated CpG sites

**gene_count**: number of genes included (even partially) inside the interval

**total_TPM**: sum of the TPMs of the genes in the interval

then the MSR and some related statistics: 
**msr**, **inverted msr**, **msr ecdf **, **inverted msr ecdf**, **residual** (residual of the linear regression between msr and meth rate), **inverted residual**.


```{r, echo=FALSE}

```

First let's see if there are pairwise correlations between the features.

###### Basic features:

```{r, echo=FALSE}
chart.Correlation(data_table[,essentials])
```

log(TPM) is considered only for fragments with at least a gene.

###### Comparison with simple MSR statistics:
```{r, echo=FALSE}
chart.Correlation(data_table[,c(essentials, essential_msr_predictors)])
```

######Comparison with other MSR statistics:
```{r, echo=FALSE}
chart.Correlation(data_table[,c(y, fancy_msr_predictors)])
```

inverted msr vs log(tpm):
```{r echo=FALSE, warning=FALSE}
library(ggplot2)
library(ggExtra)
#plot(data_table$inverted_msr, data_table$log_tpm, pch = )
#ggplot(data_table,aes(x=inverted_msr,y=log_tpm)) + stat_binhex(bins = 20) + scale_fill_viridis_c() + geom_point(shape = '.', col = 'white')
((ggplot(data_table,aes(x=inverted_msr,y=log_tpm)) + geom_point(alpha = 0.03)))
```


#### Predicting gene presence
Check if features can predict gene presence:

```{r, echo=FALSE}
gene_mask = data_table$gene_count>0
feature = feature = data_table$`meth rate`
feature_name = "methylation rate"

binary_predictivity_hist(feature,
                         mask = gene_mask,
                         xlab = feature_name, sep_names = c("one or more genes", "no genes"),
                         colors = c(5,7), breaks = 15,
                         main = sprintf("Can %s predict the presence of a gene?", feature_name),
                         max_y = 8)

```

```{r, echo=FALSE}

feature = feature = 1000/data_table$nucleotides
feature_name = "CpG density"

binary_predictivity_hist(feature,
                         mask = gene_mask,
                         xlab = feature_name, sep_names = c("one or more genes", "no genes"),
                         colors = c(5,7), breaks = 15,
                         main = sprintf("Can %s predict the presence of a gene?", feature_name),
                         max_y = 70)

```

```{r, echo=FALSE}

feature = data_table$inverted_msr
feature_name = "inverted MSR"

binary_predictivity_hist(feature,
                         mask = gene_mask,
                         xlab = "inverted MSR", sep_names = c("one or more genes", "no genes"),
                         colors = c(5,7), breaks = 15,
                         main = sprintf("Can %s predict the presence of a gene?", feature_name),
                         max_y = 30)

#feature = data_table$inverted_msr
#feature = data_table$`ecdf(msr, density)`
#feature = data_table$`inverted ecdf(msr, density)`
```

```{r, echo=FALSE}

feature = feature = data_table$`inverted ecdf`
feature_name = "inverted ecdf"

binary_predictivity_hist(feature,
                         mask = gene_mask,
                         xlab = feature_name, sep_names = c("one or more genes", "no genes"),
                         colors = c(5,7), breaks = 15,
                         main = sprintf("Can %s predict the presence of a gene?", feature_name),
                         max_y = 15)

```




```{r, echo=FALSE}

gene_0 = data_table$gene_count==0
gene_1 = data_table$gene_count==1
gene_2 = data_table$gene_count==2
gene_many = data_table$gene_count>2

boxplot(data_table$inverted_msr[gene_0], data_table$inverted_msr[gene_1], data_table$inverted_msr[gene_2], data_table$inverted_msr[gene_many], names = c("No genes", "one gene", "two genes", "more than 2 genes"), ylab = "inverted MSR")
title("Inverted MSR and number of genes")

```

```{r, echo=FALSE}
# DATASET

model_data = data_table[complete.cases(data_table[,msr_predictors]),]
cat("missing data: ", (1-(length(model_data[,1])/length(data_table[,1])))*100, "%")

# UNDERSAMPLING
train_prop = 0.2; cat("\ntrain_data_proportion: ", train_prop)
random = T

l = length(model_data[,1])
train_length = train_prop*l

if(random) 
{
  train_rows = sample(1:l, size = train_length, replace = F)
} else train_rows = 1:train_length

test_rows = (1:l)[-train_rows]

train_model_data = model_data[train_rows,]
test_model_data = model_data[test_rows,]
```

The fraction of fragments that have at least one gene inside is (train data)
```{r, echo=FALSE}
mean(train_model_data$gene_count>0, na.rm=T)
```

Logistic Regression Model for gene presence with basic predictors (nucleotides, CpG_density, meth rate, msr_density):

```{r, echo=FALSE}

# Logistics Regression
glm.fit <- glm(train_model_data$gene_count>0 ~ nucleotides+CpG_density+`meth rate`+ msr_density , data = train_model_data, family = binomial)


#summary(glm.fit)
cat("TRAIN DATA ")
nd <- train_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$gene_count>0
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2], "\n\n")

cat("TEST DATA ")
nd <- test_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$gene_count>0
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2])
```


Logistic Regression Model with inverted_msr as predictor
```{r, echo=FALSE}
glm.fit <- glm(train_model_data$gene_count>0 ~ inverted_msr , data = train_model_data, family = binomial)

#summary(glm.fit)
cat("TRAIN DATA")
nd <- train_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$gene_count>0
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2], "\n\n")

cat("TEST DATA")
nd <- test_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$gene_count>0
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2])

```


Logistic Regression Model with all predictors
```{r, echo=FALSE}
glm.fit <- glm(train_model_data$gene_count>0 ~ . , data = train_model_data[, c(basic_predictors, msr_predictors, "gene_count")], family = binomial)

#summary(glm.fit)
cat("TRAIN DATA")
nd <- train_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$gene_count>0
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2], "\n\n")

cat("TEST DATA")
nd <- test_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$gene_count>0
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2])

```




#### Predicting log(TPM)

Distribution of TPM values (only for regions that contains some genes)
```{r, echo=FALSE}
hist(log(model_data[model_data$gene_count>0, ]$total_TPM+1e-3, base = 10), xlab = "log(TPM + 1e-3)", main = "", breaks = 50)
```



Linear model for TPM with standard predictors:
```{r, echo=FALSE}
train_model_data = train_model_data[train_model_data$gene_count>0, ]
test_model_data = test_model_data[test_model_data$gene_count>0, ]

standard_predictors = c(essentials, "genes_nucleotides_count")
model = lm(log_tpm ~ . , data = train_model_data[, standard_predictors])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```

Linear model for TPM with all features and MSR statistics:
```{r, echo=FALSE}
model = lm(log_tpm ~ . , data = train_model_data[, c(standard_predictors, msr_predictors)])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```


Linear model for TPM with some features:

```{r, echo=FALSE}
model = lm(log_tpm ~ ., data = train_model_data[, c(standard_predictors, "inverted_msr")])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```


Linear model for TPM with basic features but no information about genes:
```{r, echo=FALSE}
train_model_data = model_data[train_rows,]
test_model_data = model_data[test_rows,]
train_model_data$log_tpm <- log(train_model_data$total_TPM+1e-3)
test_model_data$log_tpm <- log(test_model_data$total_TPM+1e-3)

model = lm(log_tpm ~ ., data = train_model_data[, c("log_tpm", basic_predictors)])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```


Linear model for TPM with all features but no information about genes:
```{r, echo=FALSE}
model = lm(log_tpm ~ ., data = train_model_data[, c("log_tpm", basic_predictors, msr_predictors)])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```

Linear model for TPM with inverted msr but no information about genes:
```{r, echo=FALSE}
model = lm(log_tpm ~ ., data = train_model_data[, c("log_tpm", basic_predictors, "inverted_msr")])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```
