---
title: "Tissues data comparison"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#knitr::opts_chunk$set(fig.width=20, fig.height=12) 

#setwd(dir = "Scrivania/Tesi/Methylation-project/Code/")
#source("Methylation-project/Code/expression_and_msr_experiment.R", chdir = T)
```



```{r, include=FALSE}
library("PerformanceAnalytics")
setwd(dir = "../../../")
source("Methylation-project/Code/WGBS_analysis_functions.R", chdir = T)
genebody <- readRDS("Rexperiments/genebody_improved.Rda")

names <- c("left_ventricle", "ovary", "stomach", "lung", "gm23248", "H1", "K562")

rna_files <- c("MethylationCode/MethylationData/rna-seq/ENCFF860DPP_left_ventricle_male_34.tsv",
              "MethylationCode/MethylationData/rna-seq/ENCFF061BJL_ovary_53_gene_quantifications.tsv",
              "MethylationCode/MethylationData/rna-seq/ENCFF918KPC_stomach.tsv",
              "MethylationCode/MethylationData/rna-seq/ENCFF947WLV_lung_30_gene_quantifications.tsv",
              "MethylationCode/MethylationData/rna-seq/ENCFF775DYT_gm23248_polyA.tsv",
              "MethylationCode/MethylationData/rna-seq/ENCFF653XHG_H1.tsv",
              "MethylationCode/MethylationData/rna-seq/K562.tsv"
              )

wgbs_files <- c("MethylationCode/MethylationData/wgbs/ENCFF536RSX_left_ventricle_male_34.rda",
                "MethylationCode/MethylationData/wgbs/ENCFF303ZGP_ovary_53_female.rda",
                "MethylationCode/MethylationData/wgbs/stomach.Rda",
                "MethylationCode/MethylationData/wgbs/ENCFF039JFT_lung_30_female.rda",
                "MethylationCode/MethylationData/wgbs/ENCFF752NXS_GM23248.rda",
                "MethylationCode/MethylationData/wgbs/H1.Rda",
                "MethylationCode/MethylationData/wgbs/K562.Rda")
```


```{r, include=FALSE}
setwd(dir = "../../../")
rna_data <- lapply(rna_files, read_rna_file)


rna_data <- lapply(rna_data, function(x) {x[x$gene_id %in% genebody$id]})
rna_data <- lapply(rna_data, function(x) {x[!duplicated(x$gene_id), ]})

g = rna_data[[1]]$gene_id
for(x in rna_data)
{
  g = intersect(g, x$gene_id)
}

rna_data <- lapply(rna_data, function(x) {x[x$gene_id %in% g,TPM]})
#rna_data <- lapply(rna_data, function(x) {x$TPM})
rna_data <- data.frame(rna_data)
colnames(rna_data) <- names
rna_data <- log(rna_data+1e-3)
gc()
#########################
```

Correlation in gene expression ( log(TPM+1e-3) ): 
```{r, echo=FALSE}
corrplot(cor(na.omit(rna_data)),  method = "number")
```


```{r, include=FALSE}
setwd(dir = "../../../")
wgbs_data <- lapply(wgbs_files, function(x){sum_strands(readRDS(x))})
wgbs_data <- lapply(wgbs_data, function(x){(x$prop)})
wgbs_data <- data.frame(wgbs_data)
colnames(wgbs_data) <- names
gc()
#########################
```

Methylation Levels: 
```{r, echo=FALSE}
cat("Methylation Levels")
for(i in 1:length(names))
{
  cat(names[i], ":", mean(wgbs_data[,names[i]], na.rm=T), "%\n")
}

```

Correlation in CpG methylation:
```{r, echo=FALSE}
corrplot(cor(na.omit(wgbs_data)),  method = "number")
```

Normalized Hamming distance:
```{r, echo=FALSE}
n <- 1e4
rows <- sample((1:length(wgbs_data[,1]))[complete.cases(wgbs_data)], n, replace = F)
f <- function(a,b) {mean(abs(a-b), na.rm = T)}
pairwise.plot(wgbs_data[rows, ]>=50, f)*100
```