---
title: "MSR and expression for H1, windows of 1000 sites"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#knitr::opts_chunk$set(fig.width=20, fig.height=12) 

#setwd(dir = "Scrivania/Tesi/Methylation-project/Code/")
#source("Methylation-project/Code/expression_and_msr_experiment.R", chdir = T)
```



```{r, include=FALSE}
library("PerformanceAnalytics")
setwd(dir = "../../../")
source("Methylation-project/Code/WGBS_analysis_functions.R", chdir = T)


##########################
load("Rexperiments/H1_fragments_table_stoc_V2.Rdata")
load("Rexperiments/H1_fragments_table_discrete_V2.Rdata")
load("Rexperiments/H1_fragments_table_adaptive_V2.Rdata")

genebody = readRDS("Rexperiments/genebody_improved.Rda")
#########################

y <- c("gene_count", "log_tpm")
basic_predictors <- c("nucleotides", "CpG_density", "meth rate")
essentials <- c(y, basic_predictors)
essential_msr_predictors <- c("msr", "inverted_msr")
fancy_msr_predictors <- c("ecdf", "inverted ecdf", "residual", "inverted_residual")
msr_predictors <- c(essential_msr_predictors, fancy_msr_predictors)
```


```{r, include=FALSE}
setwd(dir = "../../../")
rna <- read_rna_file("MethylationCode/MethylationData/rna-seq/ENCFF653XHG_H1.tsv")
data <- sum_strands(readRDS("MethylationCode/MethylationData/wgbs/H1.Rda"))
msr_fragments_table = fragments_table
rna_fragment_tables = readRDS("Rexperiments/H1_rna_fragment_tables.Rda")
wsize = 1
#########################
```

Here I will investigate if there is a relationship between the presence of genes and genes expression in a certain genomic region and the MSR (with some of its derivate statistics).

I chose H1 cells data in order to do this, and CpG windows of size 1000, that corresponds to a variable window size in term of nucleotides (on average about 100,000).


This is an example of total-rna-seq file, that shows for each "gene" its transcripts and some measures of expression. In this case I just kept two colums.
The first one indicates the "gene"", the second one is the Transcript Per Million that is a relative measure of how much a gene is expressed.

```{r, echo=FALSE}
#H1_rna
rna[nchar(gene_id)>6]
```

This is the annotation file that store the position occupied by each human gene.

```{r, echo=FALSE}
genebody[,c(1,2,3,4,5,6)]
```
The number of genes is much less than the ones in the total-rna-seq file, since the first one also contains so called pseudogenes and other stuff.



So the final dataFrame is the following (excluding some columns for readability):
```{r, echo=FALSE}
data_table <- join_rna_and_msr_tables(rna_fragment_tables, msr_fragments_table, wsize)

data_table <- exclude_outliers(data_table, lim = 2.5e5)
data_table$log_tpm[data_table$gene_count==0] = NA
#data_table <- exclude_outliers(data_table, lim = 2e6)
head(data_table[complete.cases(data_table), c(1,2,3,5,7,11)])
```

The full scheme includes:

**nucleotides**: number of nucleotides in the window

**CpG density**: fraction of nucleotides that is a C of a CpG site (= 1000/nucleotides)

**meth rate**: ratio of methylated CpG sites

**gene_count**: number of genes included (even partially) inside the interval

**total_TPM**: sum of the TPMs of the genes in the interval

then the MSR and some related statistics: 
**msr**, **inverted msr**, **msr ecdf **, **inverted msr ecdf**, **residual** (residual of the linear regression between msr and meth rate), **inverted residual**.



First let's see if there are pairwise correlations between the features.

###### Basic features:

```{r, echo=FALSE}
chart.Correlation(data_table[,essentials])
```

log(TPM) is considered only for fragments with at least a gene.

###### Comparison with simple MSR statistics:
```{r, echo=FALSE}
chart.Correlation(data_table[,c(essentials, essential_msr_predictors)])
```

######Comparison with other MSR statistics:
```{r, echo=FALSE}
chart.Correlation(data_table[,c(y, fancy_msr_predictors)])
```

inverted msr vs log(tpm):
```{r echo=FALSE, warning=FALSE}
library(ggplot2)
library(ggExtra)
#plot(data_table$inverted_msr, data_table$log_tpm, pch = )
#ggplot(data_table,aes(x=inverted_msr,y=log_tpm)) + stat_binhex(bins = 20) + scale_fill_viridis_c() + geom_point(shape = '.', col = 'white')
((ggplot(data_table,aes(x=inverted_msr,y=log_tpm)) + geom_point(alpha = 0.03)))
```


#### Predicting gene presence
Check if features can predict gene presence:

```{r, echo=FALSE}
gene_mask = data_table$gene_count>0
feature = feature = data_table$`meth rate`
feature_name = "methylation rate"

binary_predictivity_hist(feature,
                         mask = gene_mask,
                         xlab = feature_name, sep_names = c("one or more genes", "no genes"),
                         colors = c(5,7), breaks = 15,
                         main = sprintf("Can %s predict the presence of a gene?", feature_name),
                         max_y = 8)

```

```{r, echo=FALSE}

feature = feature = 1000/data_table$nucleotides
feature_name = "CpG density"

binary_predictivity_hist(feature,
                         mask = gene_mask,
                         xlab = feature_name, sep_names = c("one or more genes", "no genes"),
                         colors = c(5,7), breaks = 15,
                         main = sprintf("Can %s predict the presence of a gene?", feature_name),
                         max_y = 70)

```

```{r, echo=FALSE}

feature = data_table$inverted_msr
feature_name = "inverted MSR"

binary_predictivity_hist(feature,
                         mask = gene_mask,
                         xlab = "inverted MSR", sep_names = c("one or more genes", "no genes"),
                         colors = c(5,7), breaks = 15,
                         main = sprintf("Can %s predict the presence of a gene?", feature_name),
                         max_y = 30)

#feature = data_table$inverted_msr
#feature = data_table$`ecdf(msr, density)`
#feature = data_table$`inverted ecdf(msr, density)`
```

```{r, echo=FALSE}

feature = feature = data_table$`inverted ecdf`
feature_name = "inverted ecdf"

binary_predictivity_hist(feature,
                         mask = gene_mask,
                         xlab = feature_name, sep_names = c("one or more genes", "no genes"),
                         colors = c(5,7), breaks = 15,
                         main = sprintf("Can %s predict the presence of a gene?", feature_name),
                         max_y = 13)

```




```{r, echo=FALSE}

gene_0 = data_table$gene_count==0
gene_1 = data_table$gene_count==1
gene_2 = data_table$gene_count==2
gene_many = data_table$gene_count>2

boxplot(data_table$inverted_msr[gene_0], data_table$inverted_msr[gene_1], data_table$inverted_msr[gene_2], data_table$inverted_msr[gene_many], names = c("No genes", "one gene", "two genes", "more than 2 genes"), ylab = "inverted MSR")
title("Inverted MSR and number of genes")

```


The fraction of fragments that have at least one gene inside is
```{r, echo=FALSE}
mean(data_table$gene_count>0, na.rm=T)
```

Logistic Regression Model for gene presence with basic predictors (nucleotides, CpG_density, meth rate):

```{r, echo=FALSE}

# Logistics Regression
model_data = (data_table[, c("nucleotides", "CpG_density", "meth rate", "gene_count")])
model_data = model_data[complete.cases(model_data),]
glm.fit <- glm(model_data$gene_count>0 ~ nucleotides+CpG_density+`meth rate`, data = model_data, family = binomial)


#summary(glm.fit)

probabilities = predict(glm.fit, type="response")
prediction = probabilities>0.5
actual = glm.fit$data$gene_count>0
t = table(actual, prediction)
prop.table(t)

cat("\naccuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2])
```


Logistic Regression Model with inverted_msr as predictor
```{r, echo=FALSE}
model_data = (data_table[, c("inverted_msr", "gene_count")])
model_data = model_data[complete.cases(model_data),]
glm.fit <- glm(model_data$gene_count>0 ~ model_data$inverted_msr , data = model_data, family = binomial)

probabilities = predict(glm.fit, type="response")
prediction = probabilities>0.5
actual = model_data$gene_count>0
t = table(actual, prediction)
prop.table(t)

cat("\naccuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2])

```


Adding other predictors doesn't significantly improve the accuracy.


#### Predicting log(TPM)

Distribution of TPM values (only for regions that contains some genes)
```{r, echo=FALSE}

hist(log(data_table[data_table$gene_count>0, ]$total_TPM+1e-3, base = 10), xlab = "log(TPM + 1e-3)", main = "", breaks = 50)
```



Linear model for TPM with standard predictors:
```{r, echo=FALSE}
gene_presence = data_table$gene_count>0
model_data = (data_table[gene_presence,])
model_data = model_data[complete.cases(model_data),]

model = lm(log_tpm ~ . , data = model_data[, c(essentials)])
summary(model)

```

Linear model for TPM with all features and MSR statistics:
```{r, echo=FALSE}
model = lm(log_tpm ~ . , data = model_data[, c(essentials, msr_predictors, "genes_nucleotides_count")])
summary(model)
```


Linear model for TPM with some features:

```{r, echo=FALSE}

meth_rate = model_data$`meth rate`
model = lm(log_tpm ~ (model_data$inverted_msr) + (model_data$CpG_density) + (meth_rate) + (model_data$gene_count), data = model_data)
summary(model)
```





