---
title: "MSR and expression, for stomach cells and window size of 10,000"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#knitr::opts_chunk$set(fig.width=20, fig.height=12) 

#setwd(dir = "Scrivania/Tesi/Methylation-project/Code/")
#source("Methylation-project/Code/expression_and_msr_experiment.R", chdir = T)
```



```{r, include=FALSE}
library("PerformanceAnalytics")
setwd(dir = "../../../")
source("Methylation-project/Code/WGBS_analysis_functions.R", chdir = T)


##########################


#load("Rexperiments/H1_fragments_table_stoc_V2.Rdata")
#load("Rexperiments/H1_fragments_table_discrete_V2.Rdata")
#load("Rexperiments/H1_fragments_table_adaptive_V2.Rdata")

#load("Rexperiments/K562_fragments_table_stoc_V2.Rdata")
#load("Rexperiments/K562_fragments_table_discrete_V2.Rdata")
#load("Rexperiments/K562_fragments_table_adaptive_V2.Rdata")

#load("Rexperiments/stomach_fragments_table_stoc_V2.Rdata")
load("Rexperiments/stomach_fragments_table_discrete_V2.Rdata")
#load("Rexperiments/stomach_fragments_table_adaptive_V2.Rdata")

genebody = readRDS("Rexperiments/genebody_improved.Rda")
#########################

y <- c("gene_count", "log_tpm")
basic_predictors <- c("nucleotides", "CpG_density", "meth rate")
essentials <- c(y, basic_predictors)
essential_msr_predictors <- c("msr", "inverted_msr")
fancy_msr_predictors <- c("ecdf", "inverted ecdf", "residual", "inverted_residual")
msr_predictors <- c(essential_msr_predictors, fancy_msr_predictors)
```


```{r, include=FALSE}
setwd(dir = "../../../")
rna <- read_rna_file("MethylationCode/MethylationData/rna-seq/ENCFF918KPC_stomach.tsv")
#data <- sum_strands(readRDS("MethylationCode/MethylationData/wgbs/stomach.Rda"))
msr_fragments_table = stomach_fragments_table_discrete
rna_fragment_tables = readRDS("Rexperiments/stomach_rna_fragment_tables.Rda")
wsize = 2
#########################
```

Here I will investigate if there is a relationship between the presence of genes and genes expression in a certain genomic region and the MSR (with some of its derivate statistics).

I chose stomach cells data in order to do this, and CpG windows of size 10,000, that corresponds to a variable window size in term of nucleotides (on average about 1,000,000).


This is an example of total-rna-seq file, that shows for each "gene" its transcripts and some measures of expression. In this case I just kept two colums.
The first one indicates the "gene"", the second one is the Transcript Per Million that is a relative measure of how much a gene is expressed.

```{r, echo=FALSE}
#stomach_rna
rna[nchar(gene_id)>6]
```

This is the annotation file that store the position occupied by each human gene.

```{r, echo=FALSE}
genebody[,c(1,2,3,4,5,6)]
```
The number of genes is much less than the ones in the total-rna-seq file, since the first one also contains so called pseudogenes and other stuff.



So the final dataFrame is the following (excluding some columns for readability):
```{r include=FALSE}
data_table <- join_rna_and_msr_tables(rna_fragment_tables, msr_fragments_table, wsize)

#data_table <- exclude_outliers(data_table, lim = 2.5e5)
data_table$log_tpm[data_table$gene_count==0] = NA
data_table$log_std_tpm[is.infinite(data_table$log_std_tpm)] = NA

#data_table$log_std_tpm = data_table$log_std_tpm + log(data_table$gene_count-1) - log(data_table$gene_count)

data_table[data_table$gene_count>0, ]$std_TPM = data_table[data_table$gene_count>0, ]$std_TPM * sqrt((data_table[data_table$gene_count>0, ]$gene_count-1)/ (data_table[data_table$gene_count>0, ]$gene_count))

data_table <- exclude_outliers(data_table, lim = 2.5e6)
```

```{r, echo=FALSE}
head(data_table[complete.cases(data_table), c(1,2,3,5,7,11)])
```

The full scheme includes:

**nucleotides**: number of nucleotides in the window

**CpG density**: fraction of nucleotides that is a C of a CpG site (= 10000/nucleotides)

**meth rate**: ratio of methylated CpG sites

**gene_count**: number of genes included (even partially) inside the interval

**total_TPM**: sum of the TPMs of the genes in the interval

then the MSR and some related statistics: 
**msr**, **inverted msr**, **msr ecdf **, **inverted msr ecdf**, **residual** (residual of the linear regression between msr and meth rate), **inverted residual**.



First let's see if there are pairwise correlations between the features.

###### Basic features:

```{r, echo=FALSE}
chart.Correlation(data_table[,essentials])
#mi.plot(data_table[,essentials])
```

log(TPM) is considered only for fragments with at least a gene.

###### Comparison with simple MSR statistics:
```{r, echo=FALSE}
chart.Correlation(data_table[,c(essentials, essential_msr_predictors, "msr_density")])
#mi.plot(data_table[,c(essentials, essential_msr_predictors)])
```

######Comparison with other MSR statistics:
```{r, echo=FALSE}
chart.Correlation(data_table[,c(y, fancy_msr_predictors)])
#mi.plot(data_table[,c(y, fancy_msr_predictors)])
```


```{r echo=FALSE, warning=FALSE}
library(ggplot2)
library(ggExtra)
#plot(data_table$inverted_msr, data_table$log_tpm, pch = )
#ggplot(data_table,aes(x=inverted_msr,y=log_tpm)) + stat_binhex(bins = 20) + scale_fill_viridis_c() + geom_point(shape = '.', col = 'white')
(ggplot(data_table,aes(x=inverted_msr,y=log_tpm))) + geom_point(alpha = 0.1) + ggtitle("inverted msr vs log(tpm)")
```

The correlation between the total TPM with the standard deviation of the TPM is:
```{r echo=FALSE, warning=FALSE}
cor.test(data_table$total_TPM, data_table$std_TPM)$estimate
```

```{r echo=FALSE, warning=FALSE}
((ggplot(data_table,aes(x=log_tpm,y=log_std_tpm), xlim=c(-5, 5), ylim = c(-5, 5)) + geom_point(alpha = 0.1))) + geom_abline(col = 2 ) + ggtitle("log(tpm) msr vs log(std(tpm))")
```


```{r echo=FALSE, warning=FALSE}
(ggplot(data_table,aes(x=inverted_msr,y=gene_count)) + geom_point(alpha = 0.08, col = "dark blue")) +ggtitle("inverted msr vs gene number")
```


#### Predicting gene number

```{r, echo=FALSE}
hist(data_table$gene_count, main = sprintf("number of genes in a fragment, mean: %s", round(mean(data_table$gene_count,na.rm = T),2)), xlab = "genes count")
```



Negative binomial for gene number with basic predictors (nucleotides, CpG_density, meth rate, msr_density):

```{r, echo=FALSE}
library("MASS")

# Logistics Regression


model_data = data_table[complete.cases(data_table[,msr_predictors]),]
model_data = (model_data[, c("nucleotides", "CpG_density", "meth rate", "gene_count", "msr_density")])

glm.fit <- glm.nb(model_data$gene_count ~ nucleotides+CpG_density+`meth rate`+msr_density, data = model_data)


summary(glm.fit)

probabilities = predict(glm.fit, type="response")
prediction = (probabilities)
actual = model_data$gene_count
#plot(prediction-actual, col = alpha(1, 0.2))

```


Negative binomial regression Model with inverted_msr as predictor
```{r, echo=FALSE}
model_data = data_table[complete.cases(data_table[,msr_predictors]),]
model_data = (model_data[, c("inverted_msr", "gene_count")])

glm.fit <- glm.nb(model_data$gene_count ~ model_data$inverted_msr , data = model_data)

summary(glm.fit)

probabilities = predict(glm.fit, type="response")
prediction = (probabilities)
actual = model_data$gene_count
#plot(prediction-actual, col = alpha(1, 0.2))

```


Negative Binomial Regression Model with several predictors
```{r, echo=FALSE}
model_data = data_table[complete.cases(data_table[,msr_predictors]),]
model_data = (model_data[, c("gene_count", basic_predictors, msr_predictors)])
model_data = model_data[complete.cases(model_data),]
glm.fit <- glm.nb(gene_count ~ . , data = model_data)

summary(glm.fit)

probabilities = predict(glm.fit, type="response")
probabilities = (probabilities)
actual = model_data$gene_count

hist(abs(prediction-actual), main = "Error distribution compared with data", breaks = 50, xlab= "prediction-actual", probability = T, col = alpha(5, 0.7))
hist(actual, main = "error distribution", breaks = 50, xlab= "gene_count", col = alpha(2, 0.5), probability = T, add=T)

#hist(prediction-actual, col = alpha(1, 0.4), main = "error distribution", breaks = 50, xlab= "prediction-actual")
colors = c(alpha(5, 0.5), alpha(2, 0.5))
legend("right", col = colors, legend = c("prediction error","gene count"), fill = colors)
```


#### Predicting log(TPM)

Distribution of TPM values (only for regions that contains some genes).
```{r, echo=FALSE}

hist(data_table[data_table$gene_count>0,]$log_tpm, xlab = "log(TPM)", main = "", breaks = 50)
```



Linear model for log(TPM) with standard predictors:
```{r, echo=FALSE}
to_predict = "log_tpm"
gene_presence = data_table$gene_count>0
standard_predictors <- c(to_predict, basic_predictors, "gene_count", "genes_nucleotides_count", "msr_density")
model_data = data_table[complete.cases(data_table[,msr_predictors]),]
model_data = (model_data[gene_presence,standard_predictors])

#model_data = model_data[complete.cases(model_data),]

model = lm(log_tpm ~ . , data = model_data)
summary(model)

```

Linear model for TPM with all features and MSR statistics:
```{r, echo=FALSE}
model_data = data_table[complete.cases(data_table[,msr_predictors]),]
model_data = (model_data[gene_presence, c(standard_predictors, msr_predictors)])
model = lm(log_tpm ~ . , model_data)
summary(model)

hist((model$fitted.values-model$model$log_tpm), main = "Error distribution compared with centered data", breaks = 100, xlab= "prediction-actual", probability = F, col = alpha(5, 1))
hist(model$model$log_tpm-mean(model$model$log_tpm), breaks = 50, xlab= "log(tpm)", col = alpha(2, 0.4), probability = F, add=T)

colors = c(alpha(5, 0.5), alpha(2, 0.5))
legend("right", col = colors, legend = c("prediction error","centered log(tpm)"), fill = colors)
```

Linear model for TPM with all features and MSR statistics, without information about genes:
```{r, echo=FALSE}
model = lm(log_tpm ~ . , model_data[, c(to_predict, basic_predictors, msr_predictors, "msr_density")])
summary(model)
```


Linear model for TPM with some features:

```{r, echo=FALSE}

model_data = data_table[complete.cases(data_table[,msr_predictors]),]
model = lm(log_tpm ~ ., data = model_data[,c(standard_predictors, "inverted_msr")])
summary(model)
```



