---
title: "MSR and expression for H1, windows of 1000 sites"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



```{r, include=FALSE}
library("PerformanceAnalytics")
source("../../../Methylation-project/Code/WGBS_analysis_functions.R", chdir = T)


y <- c("gene_count", "log_tpm")
basic_predictors <- c("nucleotides", "CpG_density", "meth rate", "meth_autocorrelation")
essentials <- c(y, basic_predictors)
essential_msr_predictors <- c("msr", "inverted_msr")
fancy_msr_predictors <- c("ecdf", "inverted ecdf", "residual", "inverted_residual")
msr_predictors <- c(essential_msr_predictors, fancy_msr_predictors)
standard_predictors = c(essentials, "genes_nucleotides_count")
all_features <- c(standard_predictors, msr_predictors)
```

```{r, include=FALSE}
wgbs1 <- sum_strands(readRDS("../../../MethylationCode/MethylationData/wgbs/stomach.Rda")); gc()
msr_fragments_table1 = readRDS("../../../Rexperiments/stomach_msr_table_1000.Rda")
rna_fragment_table1 = readRDS("../../../Rexperiments/stomach_rna_table_1000_extended.Rda")

wgbs2 <- sum_strands(readRDS("../../../MethylationCode/MethylationData/wgbs/ENCFF039JFT_lung_30_female.rda")); gc()
msr_fragments_table2 = readRDS("../../../Rexperiments/ENCFF039JFT_lung_msr_table_1000.Rda")
rna_fragment_table2 = readRDS("../../../Rexperiments/ENCFF039JFT_lung_rna_table_1000_extended.Rda")
```


So the final dataFrame is the following (excluding some columns for readability):
```{r, echo=FALSE}
data_table1 <- join_rna_and_msr_table(rna_fragment_table1, msr_fragments_table1)
data_table2 <- join_rna_and_msr_table(rna_fragment_table2, msr_fragments_table2)

invariant_columns = c(1,2,3,4,5,6,9,19)
data_table1$meth_autocorrelation <- meth_autoc(wgbs1, data_table1)
data_table2$meth_autocorrelation <- meth_autoc(wgbs2, data_table2)

data_table <- data_table1
data_table[,-invariant_columns] <- data_table1[ ,-invariant_columns] - data_table2[ ,-invariant_columns]


data_table <- exclude_outliers(data_table, lim = 2.5e5)

```


First let's see if there are pairwise correlations between the features.

###### Basic features:

```{r, echo=FALSE}
rows <- sample(1:length(data_table[,1]), size = 4e3)
chart.Correlation(data_table[rows,essentials])
```

log(TPM) is considered only for fragments with at least a gene.

###### Comparison with simple MSR statistics:
```{r, echo=FALSE}
chart.Correlation(data_table[rows,c(essentials, essential_msr_predictors)])
```

######Comparison with other MSR statistics:
```{r, echo=FALSE}
chart.Correlation(data_table[rows,c(y, fancy_msr_predictors)])
```




inverted msr vs log(tpm):
```{r, echo=FALSE, warning=FALSE}
library(ggplot2)
library(ggExtra)
#plot(data_table$inverted_msr, data_table$log_tpm, pch = )
#ggplot(data_table,aes(x=inverted_msr,y=log_tpm)) + stat_binhex(bins = 20) + scale_fill_viridis_c() + geom_point(shape = '.', col = 'white')

((ggplot(data_table,aes(x=inverted_msr,y=log_tpm)) + geom_point(alpha = 0.03))) + ylab("log (TPM)")
```


```{r, echo=FALSE, warning=FALSE}
((ggplot(data_table,aes(x=residual,y=log_tpm)) + geom_point(alpha = 0.03)))+ ylab("log (TPM)")
```

```{r, echo=FALSE, warning=FALSE}
((ggplot(data_table,aes(x=msr_density,y=log_tpm)) + geom_point(alpha = 0.03)))+ ylab("log (TPM)")
```

```{r, echo=FALSE}
threshold = 0

cols = c(alpha(2,0.5), alpha(3,0.5))
hist(data_table[data_table$log_tpm<=threshold,]$inverted_msr, breaks = 40, col = cols[1], probability = F, xlab = "inverted msr", main = "Distribution comparison")
hist(data_table[data_table$log_tpm>threshold,]$inverted_msr, breaks = 40, col = cols[2], add = T, probability = F)

legend("top", legend=c(paste("TPM <=",threshold), c(paste("TPM >",threshold))), col=cols, fill = cols)
```

```{r, echo=FALSE}
threshold = 0

cols = c(alpha(2,0.5), alpha(3,0.5))
hist(data_table[data_table$log_tpm<=threshold,]$msr_density, breaks = 40, col = cols[1], probability = F, xlab = "discretized meth rate", main = "Distribution comparison", ylim = c(0,3000))
hist(data_table[data_table$log_tpm>threshold,]$msr_density, breaks = 40, col = cols[2], add = T, probability = F, ylim = c(0,3000))

legend("top", legend=c(paste("TPM <=",threshold), c(paste("TPM >",threshold))), col=cols, fill = cols)
```

```{r, echo=FALSE}
# DATASET

model_data = data_table[complete.cases(data_table[,msr_predictors]),]
cat("missing data: ", (1-(length(model_data[,1])/length(data_table[,1])))*100, "%")

# UNDERSAMPLING
train_prop = 0.2; cat("\ntrain_data_proportion: ", train_prop)
random = T

l = length(model_data[,1])
train_length = train_prop*l

if(random) 
{
  train_rows = sample(1:l, size = train_length, replace = F)
} else train_rows = 1:train_length

test_rows = (1:l)[-train_rows]

train_model_data = model_data[train_rows,]
test_model_data = model_data[test_rows,]
```

The fraction of fragments that have TPM>0 (train data)
```{r, echo=FALSE}
mean(train_model_data$total_TPM>threshold, na.rm=T)
```

Logistic Regression Model for TPM presence with basic predictors (nucleotides, CpG_density, meth rate, msr_density):

```{r, echo=FALSE}

# Logistics Regression
glm.fit <- glm(train_model_data$total_TPM>threshold ~ nucleotides+CpG_density+`meth rate`+ msr_density , data = train_model_data, family = binomial)


#summary(glm.fit)
cat("TRAIN DATA ")
nd <- train_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$total_TPM>threshold
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2], "\n\n")

cat("TEST DATA ")
nd <- test_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$total_TPM>threshold
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2])
```


Logistic Regression Model with inverted_msr as predictor
```{r, echo=FALSE}
glm.fit <- glm(train_model_data$total_TPM>threshold ~ inverted_msr , data = train_model_data, family = binomial)

#summary(glm.fit)
cat("TRAIN DATA")
nd <- train_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$total_TPM>threshold
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2], "\n\n")

cat("TEST DATA")
nd <- test_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$total_TPM>threshold
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2])

```


Logistic Regression Model with all predictors
```{r, echo=FALSE}
glm.fit <- glm(train_model_data$total_TPM>threshold ~ . , data = train_model_data[, c(basic_predictors, msr_predictors)], family = binomial)

#summary(glm.fit)
cat("TRAIN DATA")
nd <- train_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$total_TPM>threshold
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2], "\n\n")

cat("TEST DATA")
nd <- test_model_data
probabilities = predict(glm.fit, newdata = nd, type="response")
prediction = probabilities>0.5
actual = nd$total_TPM>threshold
t = table(actual, prediction)
#prop.table(t)

cat("accuracy: ", prop.table(t)[1,1]+prop.table(t)[2,2])

```



#### Predicting log(TPM)
Distribution of TPM values (only for regions that contains some genes)
```{r, echo=FALSE}
hist(log(model_data[model_data$gene_count>0, ]$total_TPM+1e-3, base = 10), xlab = "log(TPM + 1e-3)", main = "", breaks = 50)
```



Linear model for TPM with standard predictors, available information about genes:
```{r, echo=FALSE}
model = lm(log_tpm ~ . , data = train_model_data[, standard_predictors])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```

Linear model for TPM with all features and MSR statistics, available information about genes:
```{r, echo=FALSE}
model = lm(log_tpm ~ . , data = train_model_data[, c(standard_predictors, msr_predictors)])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```


Linear model for TPM with adding inverted msr, available information about genes:

```{r, echo=FALSE}
model = lm(log_tpm ~ ., data = train_model_data[, c(standard_predictors, "inverted_msr")])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```


Linear model for TPM with basic features but no information about genes:
```{r, echo=FALSE}
model = lm(log_tpm ~ ., data = train_model_data[, c("log_tpm", basic_predictors)])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```


Linear model for TPM with all features but no information about genes:
```{r, echo=FALSE}
model = lm(log_tpm ~ ., data = train_model_data[, c("log_tpm", basic_predictors, msr_predictors)])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```

Linear model for TPM with inverted msr but no information about genes:
```{r, echo=FALSE}
model = lm(log_tpm ~ ., data = train_model_data[, c("log_tpm", basic_predictors, "inverted_msr")])
summary(model)

cat("Test data R squared: ", 1-(var(predict(model, test_model_data, type="response")-test_model_data$log_tpm))/var(test_model_data$log_tpm) )
```
